### PR TITLE
Java: Gaussian confidence interval implementation

### PR DESCRIPTION
**Pull request by laviniamnedelea**

### CODE REVIEW

File: java/main/com/google/privacy/differentialprivacy/BUILD [hunk 1]
```diff
@@ -40,14 +40,14 @@ java_library(
...
                "@com_google_differential_privacy//proto:data_java_proto",
-               "@maven//:org_apache_commons_commons_math3",
+               "@com_google_differential_privacy//proto:summary_java_proto",
                "@maven//:com_google_auto_value_auto_value",
                "@maven//:com_google_auto_value_auto_value_annotations",
                "@maven//:com_google_code_findbugs_jsr305",
                "@maven//:com_google_errorprone_error_prone_annotations",
                "@maven//:com_google_guava_guava",
                "@maven//:com_google_protobuf_protobuf_java",
+               "@maven//:org_apache_commons_commons_math3"
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/BUILD:51]: Could you please redo the change here? It's not related to this PR.

File: java/main/com/google/privacy/differentialprivacy/DpPreconditions.java [hunk 2]
```diff
@@ -30,10 +30,25 @@
 
   private DpPreconditions() {}
 
+  static void checkEpsilonVeryStrict(double epsilon) {
+    checkArgument(
+            epsilon >= 1.0 / (1L << 50) && epsilon < POSITIVE_INFINITY,
+            "epsilon must be at least 2^-50 (and cannot be infinity). Provided value: %s",
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/DpPreconditions.java:36]: Shouldn't this be `%f`?

File: java/main/com/google/privacy/differentialprivacy/DpPreconditions.java [hunk 3]
```diff
@@ -30,10 +30,25 @@
...
+            epsilon);
+  }
+
+  static void checkEpsilonStrict(double epsilon) {
+    checkArgument(
+            epsilon > 0 && epsilon < POSITIVE_INFINITY,
+            "epsilon must be strictly positive (and cannot be infinity). Provided value: %s",
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/DpPreconditions.java:43]: Shouldn't this be `%f`?

File: java/main/com/google/privacy/differentialprivacy/DpPreconditions.java [hunk 4]
```diff
@@ -30,10 +30,25 @@
...
+            epsilon);
+  }
+
   static void checkEpsilon(double epsilon) {
-    checkArgument(epsilon >= 1.0 / (1L << 50)
-       && epsilon < POSITIVE_INFINITY,
-        "epsilon must be > 0 and < infinity. Provided value: %s", epsilon);
+    checkArgument(
+            epsilon >= 0 && epsilon < POSITIVE_INFINITY,
+            "epsilon must be nonnegative (and cannot be infinity). Provided value: %s",
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/DpPreconditions.java:50]: Shouldn't this be `%f`?

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 5]
```diff
@@ -138,14 +144,39 @@ public ConfidenceInterval computeConfidenceInterval(
...
+          double epsilon,
+          Double delta,
+          double alpha) {
+    CheckConfidenceIntervalParameters(l0Sensitivity, lInfSensitivity, epsilon, delta, alpha);
+    double l2Sensitivity = lInfSensitivity * Math.sqrt((double) l0Sensitivity);
+    double sigma = getSigma(l2Sensitivity, epsilon, delta);
+    ConfidenceInterval confInt = computeConfidenceInterval((double) noisedX, sigma, alpha);
+    confInt =
+            ConfidenceInterval.create(
+                    Math.round(confInt.lowerBound()), Math.round(confInt.upperBound()));
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:159]: The formatting here looks weird. Could you move `ConfidenceInterval.create(` one line above?

(You could also create the confidenceInterval object and return in the same expression like we did for Go)

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 6]
```diff
@@ -138,14 +144,39 @@ public ConfidenceInterval computeConfidenceInterval(
...
+    return confInt;
+  }
+
+  /**
+   * Returns {@link ConfidenceInterval} gaussian object with 1-{@code alpha} confidence level and
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:164]: nit: Is `gaussian` here necessary?

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 7]
```diff
@@ -138,14 +144,39 @@ public ConfidenceInterval computeConfidenceInterval(
...
+    double upperBound = noisedX - z;
+    return ConfidenceInterval.create(lowerBound, upperBound);
+  }
+
+  /**
+   * inverseCDFGaussian returns z with the property that Pr( Y <= z ) = p, where p is {@code
+   * confidenceLevel} and Y is a random variable.
+   */
+  private double inverseCDFGaussian(double sigma, double p) {
+    return -sigma * Math.sqrt(2) * Erf.erfcInv(2.00 * p);
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:179]: Would `2` work instead of `2.00`? (similar to Go)

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 8]
```diff
@@ -138,14 +144,39 @@ public ConfidenceInterval computeConfidenceInterval(
...
+                    Math.round(confInt.lowerBound()), Math.round(confInt.upperBound()));
+    return confInt;
+  }
+
+  /**
+   * Returns {@link ConfidenceInterval} gaussian object with 1-{@code alpha} confidence level and
+   * {@code sigma} parameters.
+   */
+  public ConfidenceInterval computeConfidenceInterval(double noisedX, double sigma, double alpha) {
+    double z = inverseCDFGaussian(sigma, alpha / 2);
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:168]: Please comment that this is negative (similar to Go)

File: java/main/com/google/privacy/differentialprivacy/Noise.java [hunk 9]
```diff
@@ -17,7 +17,7 @@
 package com.google.privacy.differentialprivacy;
 
 import com.google.differentialprivacy.SummaryOuterClass.MechanismType;
-import javax.annotation.Nullable;
+import jdk.internal.jline.internal.Nullable;
```
Comment (miracvbasaran) [inline java/main/com/google/privacy/differentialprivacy/Noise.java:20]: Could you please redo the change here?

File: java/tests/com/google/privacy/differentialprivacy/BUILD [hunk 10]
```diff
@@ -44,13 +44,15 @@ java_test(
...
-               "@maven//:com_google_guava_guava",
-               "@maven//:com_google_truth_truth",
-               "@maven//:junit_junit",
-               "@maven//:org_mockito_mockito_core",
-               "@maven//:com_google_protobuf_protobuf_java",
+              ":autovalue",
+              "//main/com/google/privacy/differentialprivacy",
+              "@com_google_differential_privacy//proto:summary_java_proto",
+              "@maven//:com_google_auto_value_auto_value",
+              "@maven//:com_google_auto_value_auto_value_annotations",
```
Comment (miracvbasaran) [inline java/tests/com/google/privacy/differentialprivacy/BUILD:51]: Is this addition necessary? if not, please redo.

File: java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+                            5,
+                            36,
+                            0.8,
+                            0.8,
+                            7.856382354e-10,
+                            ConfidenceInterval.create(-96.6140883158, 236.6140883158)));
+        }
+
+        @Test
+        public void computeConfidenceIntervalDouble() {
```
Comment (miracvbasaran) [inline java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java:101]: I don't think this is the proper Java unit test naming. Probably something like `computeConfidenceInterval_hasAccurateResultsForDoubles()` makes more sense.

In general though, you seem to have a class for each test, I thought it is preferable to have a single class for all tests but multiple methods for each test.

But Dennis could correct me on this.

**Review by miracvbasaran — CHANGES_REQUESTED** [review]

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java (hunk 6 above)
Comment (laviniamnedelea) [reply java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:164]: Hm, since it's in the GaussianNoise document i think you're right, it's not.

File: java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import javax.annotation.Nullable;
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.Collection;
+
+@RunWith(Enclosed.class)
+public class ConfidenceIntervalGaussianTest {
```
Comment (dkrft) [inline java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java:19]: nit: can we name this GaussianNoiseConfidenceIntervalTest, that way it will be next to GaussianNoiseTest when ordered alphabetically ;)

Also, we should probably test computeConfidenceInterval with sigma as a parameter since this is also part of the public facing API

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 13]
```diff
@@ -143,18 +149,47 @@ public ConfidenceInterval computeConfidenceInterval(
...
+        Math.round(confInt.lowerBound()), Math.round(confInt.upperBound()));
+  }
+
+  /**
+   * Returns {@link ConfidenceInterval} object with 1-{@code alpha} confidence level and
+   * {@code sigma} parameters.
+   */
+  public ConfidenceInterval computeConfidenceInterval(double noisedX, double sigma, double alpha) {
+    double z = inverseCDFGaussian(sigma, alpha / 2); // z will hold a negative value.
+    double lowerBound = noisedX + z;
```
Comment (dkrft) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:167]: super duper nit: consider putting these three lines into 1, i.e., return ConfidenceInterval.create(noisedX + z, noisedX - z)

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 14]
```diff
@@ -143,18 +149,47 @@ public ConfidenceInterval computeConfidenceInterval(
...
+    double upperBound = noisedX - z;
+    return ConfidenceInterval.create(lowerBound, upperBound);
+  }
+
+  /**
+   * inverseCDFGaussian returns z with the property that Pr( Y <= z ) = p, where p is {@code
+   * confidenceLevel} and Y is a random variable.
+   */
+  private double inverseCDFGaussian(double sigma, double p) {
```
[the remark this answers is not in the document]
Comment (dkrft) [reply java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:176]: also, per java style guide this should be "Cdf" rather than "CDF"

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 15]
```diff
@@ -143,18 +149,47 @@ public ConfidenceInterval computeConfidenceInterval(
...
+    return ConfidenceInterval.create(
+        Math.round(confInt.lowerBound()), Math.round(confInt.upperBound()));
+  }
+
+  /**
+   * Returns {@link ConfidenceInterval} object with 1-{@code alpha} confidence level and
+   * {@code sigma} parameters.
+   */
+  public ConfidenceInterval computeConfidenceInterval(double noisedX, double sigma, double alpha) {
+    double z = inverseCDFGaussian(sigma, alpha / 2); // z will hold a negative value.
```
[the remark this answers is not in the document]
Comment (miracvbasaran) [reply java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:166]: When I was reading the PR in Go, there was no comment initially, and I was confused when I read `lowerBound = noisedX +z`. I actually thought it was a bug and was about to comment on it but then looked into inverseCDFGaussian computation and saw that it was negative.

I think the comment is useful for people who are not super familiar with this type of math.

[the remark this answers is not in the document]
Comment (kaelhawaty) [reply java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:166]: I agree, I wasn't aware of this too until we discussed it. I think a comment about why we are doing it this way make sense to me.

File: java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java [hunk 16]
```diff
@@ -0,0 +1,289 @@
...
+    @RunWith(Parameterized.class)
+    public static class computeConfidenceIntervalDoubleTests {
+        private final ConfidenceIntervalTestCase testCase;
+
+        public computeConfidenceIntervalDoubleTests(ConfidenceIntervalTestCase testCase) {
+            this.testCase = testCase;
+        }
+
+        @Parameterized.Parameters(name = "{index}: = failed")
+        public static Collection<Object> getTestCases() {
```
[the remark this answers is not in the document]
Comment (kaelhawaty) [reply java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java:61]: We discussed this with Maria earlier, the main argument was we wanted to do argument checking tests similar to the ones we had in Go and those "Arbitrary" tests. This would probably take 40 or more separate function tests, a lot of repeated code and would be not as readable as this.

I think it makes sense to add those the tests you suggested, that would be even more test cases which would be fine with this approach. We can discuss this with Maria again. Also, if we chose to add them, we need to add them to the Go library to in a separate PR.

File: java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java (hunk 12 above)
Comment (kaelhawaty) [reply java/tests/com/google/privacy/differentialprivacy/ConfidenceIntervalGaussianTest.java:19]: I think "computeConfidenceInterval with sigma as a parameter" should be probably be private because it doesn't have argument checking which makes sense because it is called as a common function from both interface functions.

File: java/main/com/google/privacy/differentialprivacy/GaussianNoise.java [hunk 17]
```diff
@@ -143,9 +148,32 @@ public ConfidenceInterval computeConfidenceInterval(
...
       double epsilon,
       Double delta,
-      double confidenceLevel) {
-    // TODO: Implement confidence interval computation.
-    throw new UnsupportedOperationException("Not implemented yet.");
+      double alpha) {
+    checkConfidenceIntervalParameters(l0Sensitivity, lInfSensitivity, epsilon, delta, alpha);
+    double l2Sensitivity = Noise.getL2Sensitivity(l0Sensitivity, lInfSensitivity);
+    double sigma = getSigma(l2Sensitivity, epsilon, delta);
+    ConfidenceInterval confInt = computeConfidenceInterval(0.0, sigma, alpha);
```
Comment (MashaTelyatnikova) [inline java/main/com/google/privacy/differentialprivacy/GaussianNoise.java:155]: rename to intervalAroundZero or confIntAroundZero (like Karim did)

File: java/main/com/google/privacy/differentialprivacy/SecureNoiseMath.java [hunk 18]
```diff
@@ -85,4 +85,18 @@ public static double roundToMultipleOfPowerOfTwo(double x, double granularity) {
       return x;
     }
   }
+
+  /** Computes the closest double value that is larger than or equal to the provided long value. */
+  public static double nextLargerDouble(long n) {
```
Comment (MashaTelyatnikova) [inline java/main/com/google/privacy/differentialprivacy/SecureNoiseMath.java:90]: I believe you don't need to add this (when Karim will merge his version). We improved documentation a bit already, so better to keep his version

File: java/tests/com/google/privacy/differentialprivacy/BUILD [hunk 19]
```diff
@@ -44,13 +44,15 @@ java_test(
...
-               "@com_google_differential_privacy//proto:summary_java_proto",
-               "@maven//:com_google_guava_guava",
-               "@maven//:com_google_truth_truth",
-               "@maven//:junit_junit",
-               "@maven//:org_mockito_mockito_core",
-               "@maven//:com_google_protobuf_protobuf_java",
+              ":autovalue",
+              "//main/com/google/privacy/differentialprivacy",
+              "@com_google_differential_privacy//proto:summary_java_proto",
+              "@maven//:com_google_auto_value_auto_value",
```
Comment (MashaTelyatnikova) [inline java/tests/com/google/privacy/differentialprivacy/BUILD:50]: also will not need these changes, it will be added in Karim PR

File: java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java [hunk 20]
```diff
@@ -0,0 +1,727 @@
...
+  private static final double DEFAULT_DELTA = 0.3;
+  private static final double DEFAULT_NOISED_X = 70;
+  private static final double DEFAULT_ALPHA = 0.1;
+
+  private boolean approxEqual(double a, double b) {
+    double mxMagnitude = Math.max(Math.abs(a), Math.abs(b));
+    return Math.abs(a - b) <= TOLERANCE * mxMagnitude;
+  }
+
+  private void isEqual(ConfidenceInterval a, ConfidenceInterval b) {
```
Comment (MashaTelyatnikova) [inline java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java:26]: we don't need this, we decided to use isApproxEqual for all cases.
Also rename isApproxEqual to verifyApproxEqual

File: java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java [hunk 21]
```diff
@@ -0,0 +1,727 @@
...
+        .that(approxEqual(a.lowerBound(), b.lowerBound()))
+        .isTrue();
+    assertWithMessage("Upper bounds are not equal.")
+        .that(approxEqual(a.upperBound(), b.upperBound()))
+        .isTrue();
+  }
+
+  @Test
+  public void computeConfidenceInterval_double_arbitraryTest() {
+    ConfidenceInterval confInt =
```
Comment (MashaTelyatnikova) [inline java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java:48]: confInt -> actual (everywhere) or actualConfInt and then expectedConfInt

File: java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java [hunk 22]
```diff
@@ -0,0 +1,727 @@
...
+            /* lInfSensitivity */ 36,
+            /* epsilon */ 0.8,
+            /* delta */ 0.8,
+            /* alpha */ 0.2);
+    ConfidenceInterval expected = ConfidenceInterval.create(35.26815080641682, 104.73184919358317);
+    isEqual(confInt, expected);
+  }
+
+  @Test
+  public void computeConfidenceInterval_double_highAlpha() {
```
Comment (MashaTelyatnikova) [inline java/tests/com/google/privacy/differentialprivacy/GaussianNoiseConfidenceIntervalTest.java:61]: forDouble and forLong everywhere

**Review by MashaTelyatnikova — CHANGES_REQUESTED** [review]

**Review by MashaTelyatnikova — APPROVED** [review]
